pkg/authentication/authentication.go

```diff
@@ -0,0 +1,318 @@
+}
+
+type allAuths struct {
+	Auths map[string]encodedAuth `json:"auths"`
+}
+
+var (
+	// ErrNotLoggedIn is returned for users not logged into a registry
+	// that they are trying to logout of
+	ErrNotLoggedIn = errors.Errorf("not logged in")
+)
+
+// SetAuthentication stores the base64 encoding of username and password
+// in the auth file if the credentials are authenticated
+func SetAuthentication(username, password, server string) error {
+	path, err := getPathToAuth()
+	if err != nil {
+		return errors.Wrapf(err, "error getting path to auth file")
+	}
+
+	server = getFullURL(server)
+
+	isValid, err := checkAuthentication(username, password, server)
+	if err != nil {
+		return errors.Wrapf(err, "an error occured while authenticating credentials")
+	}
+	if isValid {
+		encodedCreds := encodeUserAndPass(username, password)
+		return addAuthToJSON(server, encodedCreds, path)
+	}
+	return nil
+}
+
+// GetAuthentication returns the value of the registry url stored in
+// the auth.json file
+func GetAuthentication(server string) (string, error) {
+	server = getFullURL(server)
+
+	// Check containers-storage auth.json file first
+	path, err := getPathToAuth()
+	if err != nil {
+		return "", errors.Wrapf(err, "error getting path to auth file")
+	}
+	auths, err := readJSONFile(path)
+	if err != nil {
+		return "", errors.Wrapf(err, "error reading JSON file %q", path)
+	}
+	if val, ok := auths.Auths[server]; ok {
+		return val.Auth, nil
+	}
+
+	// check .docker/config.json as secondary option
+	path = os.Getenv("HOME") + "/.docker/config.json"
+	auths, err = readJSONFile(path)
+	if err != nil {
+		return "", errors.Wrapf(err, "error reading JSON file %q", path)
+	}
+	if val, ok := auths.Auths[server]; ok {
+		return val.Auth, nil
+	}
+
+	return "", nil
+}
+
+// RemoveAuthentication deletes the credentials stored in auth.json
+func RemoveAuthentication(server string) (string, error) {
+	server = getFullURL(server)
+	path, err := getPathToAuth()
+	if err != nil {
+		return "", errors.Wrapf(err, "error getting path to auth file")
+	}
+	auths, err := readJSONFile(path)
+	if err != nil {
+		return "", errors.Wrapf(err, "error reading JSON file %q", path)
+	}
+	if _, ok := auths.Auths[server]; ok {
+		delete(auths.Auths, server)
+	} else {
+		return "", ErrNotLoggedIn
+	}
+
+	file, err := os.OpenFile(path, os.O_RDWR, 0755)
+	if err != nil {
+		return "", errors.Wrapf(err, "error opening file %q", path)
+	}
+	defer file.Close()
+
+	newData, err := json.MarshalIndent(auths, "", "\t")
+	if err != nil {
+		return "", errors.Wrapf(err, "error marshaling JSON %q", path)
+	}
+
+	if err = ioutil.WriteFile(path, newData, 0755); err != nil {
+		return "", errors.Wrapf(err, "error writing to file %q", path)
+	}
+	return server, nil
+}
+
+// GetDockerPublicRegistryURL returns the registry url of "docker.io"
+func GetDockerPublicRegistryURL(server string) string {
+	if server == "docker.io" {
+		return "https://registry-1.docker.io"
+	}
+	return server
```

Comment:
What is using this? There is no user in this repo.

It seems weird to return a URL-like string in one case and an undecorated host-name in the other; and it also feels like the kind of internal implementation detail of the protocol that might be better to hide inside a package instead of exposing it is a very public API point.

Reply:
Yes I agree.

Reply:
@mtrmac where would you suggest I put this?

Reply:
Why is this needed?

Reply:
Since there is no default registry, and docker.io maps to registry-1.docker.io, I thought it would be good to add this in the case the user types `kpod login docker.io`. Maybe `containers/image/docker/docker_client.go` does this somewhere, not sure but I will look into it.

Reply:
Yes I would figure this happens somewhere else in the stack or in DNS.  Hard coding this could cause us issues if Docker inc in the future changes docker.io to pint at registry-foobar.docker.io.

Reply:
Sorry, I didn’t realize that both the plain-host name and URI forms are actually valid keys in `auths` (or at least I do have both in my `~/.docker/config.json`—except that I have `index.docker.io/v1/` in there, not `registry-1.docker.io`).

This mapping _is_ actually hard-coded in the clients, not in DNS, sadly.

If possible, I’d prefer this to be completely hidden from external users: let them say just `docker.io` as if it were any `example.com`, and do this mapping internally in the config file parser/editor package.